File: day06-java/src/main/java/de/birgitkratz/aoc2020/day06/CustomCustoms.java
package de.birgitkratz.aoc2020.day06;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.List;
import java.util.Objects;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.stream.Collectors;

public class CustomCustoms {

    private final Path path =
            Paths.get(Objects.requireNonNull(this.getClass().getClassLoader().getResource("input.txt")).getFile());

    int processPart1() throws IOException {
        final List<String> lines = Files.readAllLines(path);

        return groupInput1(lines).stream()
                .map(s -> s.chars()
                        .mapToObj(i -> Character.toString((char) i))
                        .distinct()
                        .collect(Collectors.joining()))
                .mapToInt(String::length)
                .sum();
    }

    int processPart2() throws IOException {
        final List<String> lines = Files.readAllLines(path);

        return groupInput2(lines).stream()
                .mapToInt(String::length)
                .sum();
    }

    List<String> groupInput1(final List<String> input) {
        List<String> groups = new ArrayList<>();
        final String[] group = {""};

        input.forEach(line -> {
            if (line.isBlank()) {
                groups.add(group[0].trim());
                group[0] = "";
            } else {
                group[0] += line.trim();
            }
        });
        groups.add(group[0].trim());
        return groups;
    }

    List<String> groupInput2(final List<String> input) {
        List<String> groups = new ArrayList<>();

        final AtomicInteger index = new AtomicInteger();
        final List<String>[] group = new List[] {new ArrayList<>()};
        input.forEach(line -> {
            if (!line.isBlank()) {
                if (index.get() == 0) {
                    group[0] = stringToListOfCharacterStrings(line);
                } else {
                    group[0].retainAll(stringToListOfCharacterStrings(line));
                }
                index.getAndIncrement();
            } else {
                groups.add(listOfCharacterStringsToString(group[0]));
                index.set(0);
                group[0] = new ArrayList<>();
            }
        });
        groups.add(listOfCharacterStringsToString(group[0]));
        return groups;
    }

    private String listOfCharacterStringsToString(final List<String> elements) {
        return String.join("", elements);
    }

    private List<String> stringToListOfCharacterStrings(final String line) {
        return line.chars()
                .mapToObj(i -> String.valueOf((char) i))
                .collect(Collectors.toList());
    }
}

File: day06-kotlin/src/main/kotlin/de/birgitkratz/aoc2020/day06/CustomCustoms.kt
package de.birgitkratz.aoc2020.day06

import java.io.File
import java.nio.charset.StandardCharsets

class CustomCustoms {

    fun processPart1(): Int {
        return File("./input/input.txt").useLines(StandardCharsets.UTF_8, ::processInput1)
    }

    fun processPart2(): Int {
        return File("./input/input.txt").useLines(StandardCharsets.UTF_8, ::processInput2)
    }

    fun processInput1(lines: Sequence<String>): Int {
        return groupInput1(lines)
            .map { s: String -> s.chunked(1).distinct().joinToString("").length }
            .sum()
    }

    fun processInput2(lines: Sequence<String>): Int {
        return groupInput2(lines)
            .map { s: String -> s.length }
            .sum()
    }

    fun groupInput1(input: Sequence<String>): Sequence<String> {
        val groups = mutableListOf<String>()
        var group = ""

        input.forEach { line ->
            when {
                line.isBlank() -> {
                    groups.add(group.trim())
                    group = ""
                }
                else -> {
                    group += line.trim()
                }
            }
        }
        groups.add(group.trim())
        return groups.asSequence()
    }

    fun groupInput2(input: Sequence<String>): Sequence<String> {
        val groups: MutableList<String> = ArrayList()
        var index = 0
        var group: MutableList<String> = ArrayList()
        input.forEach { line  ->
            when {
                line.isNotBlank() -> {
                    when (index) {
                        0    -> group = line.chunked(1) as MutableList<String>
                        else -> group.retainAll(line.chunked(1))
                    }
                    index++
                }
                else -> {
                    groups.add(group.joinToString(""))
                    index = 0
                    group = ArrayList()
                }
            }
        }
        groups.add(group.joinToString(""))
        return groups.asSequence()
    }
}